include/zephyr/logging/log.h

```diff
@@ -87,15 +97,461 @@ extern "C" {
  * @param ... A string optionally containing printk valid conversion specifier,
  * followed by as many values as specifiers.
  */
-#define LOG_WRN_ONCE(...)					\
-	do {							\
-		static uint8_t __warned;			\
-		if (unlikely(__warned == 0)) {			\
-			Z_LOG(LOG_LEVEL_WRN, __VA_ARGS__);	\
-			__warned = 1;				\
-		}						\
+#define LOG_WRN_ONCE(...)                                                                          \
+	do {                                                                                       \
+		static atomic_t __warned;                                                          \
+		if (unlikely(atomic_cas(&__warned, 0, 1))) {                                       \
+			Z_LOG(LOG_LEVEL_WRN, __VA_ARGS__);                                         \
+		}                                                                                  \
+	} while (0)
+
+/**
+ * @brief Core rate-limited logging macro for regular messages
+ *
+ * @details Internal macro that provides rate-limited logging functionality
+ * for regular log messages. Uses atomic operations to ensure thread safety
+ * in multi-threaded environments. Only one thread can successfully log a
+ * message within the specified rate limit period.
+ *
+ * @param _level Log level (LOG_LEVEL_ERR, LOG_LEVEL_WRN, etc.)
+ * @param _rate_ms Minimum interval in milliseconds between log messages.
+ * @param ... Arguments to pass to the logging function.
+ */
+#define _LOG_RATELIMIT_CORE(_level, _rate_ms, ...)                                                 \
+	do {                                                                                       \
+		static atomic_t __last_log_time;                                                   \
+		static atomic_t __skipped_count;                                                   \
+		uint32_t __now = k_uptime_get_32();                                                \
```

Comment:
that would require `#include <zephyr/kernel.h>` but I would prefer that it's not included in `log.h` as we try to include as little as possible to allow including `log.h` anywhere. Maybe extern?

Reply:
maybe a little hacky `extern uint32_t arch_k_cycle_get_32(void)`?

Reply:
Even that is static line, but I see `sys_clock_cycle_get_32` can be used, WDYT?

Reply:
But then we'd need `k_ticks_to_ms_floor64` which needs another header.